feat: Add support for custom block time

MujkicA opened the pull request:
Closes #606 

Adds an optional parameter to `produce_blocks` to set the start time and the interval of the produced blocks. 
As can be seen from the included test, querying the node to find out block times could use some UX improvements.

This PR also includes two new types that act as wrappers/anti-corruption layer for the corresponding gql-client types:

- `Block` (added to improve UX around retrieving block times)
- `TransactionResponse` (added for consistency)

The types expose a reasonable subset of the fields provided in the client types as methods. If the need should arise, we can cache the conversion of the fields instead of performing it on each method call. 

Until now the client `Block` type was not exposed by the provider, if some external code makes use of it, it does so via the client method directly. Therefore this is a safe change and won't annoy any users.
The same is not true for 'TransactionResponse' since this was returned by some provider methods and has now been swapped for our wrapper. We can of course choose postpone this change to a separate issue since it is a breaking change but at the same time not relevant for enabling custom block times.

An open question is do we also want to wrap types like `Message`, `Coin`, `ChainInfo`?

segfault-magnet reviewed (commented):
Trying to think of reasons why anyone ever wraps types:
* You can simplify it, or _facade_ it if you like. Perhaps users find the original one painful to work with, and for whatever reason, the owners of the type are reluctant to change it.
* We want to add something to it. Data or behavior our users might find pleasant to have on the type itself.
* We want to hide stuff the users shouldn't know about.
* The type changes too often. Data gets shifted around various API-s. We want to stabilize it how we see fit. If the type suddenly drops some data it used to have, we'll poll some other API and enrich it with that data again.

And probably other reasons I can't think of right now.

But, for consistency's sake and following the rule of least surprise, if we wrap one API type, we should probably wrap all so that users aren't confused. 

As far as anti-corruption properties, I'm not sure if the node stops sending some data that we can get it anywhere else. The only thing we might protect from is if they decide to split up the API and we don't want that to cause breaking changes for us.

But truth is, we're at v0.*, and we will probably continue to fit the node API like a glove until we stabilize.

It seems to me that simplification and data/behavior enrichment might be the only reasons we might have. The whole provider itself exists mostly because of the aforementioned.

Dunno. My 2 cents.

Br1ght0ne commented on packages/fuels-signers/src/provider.rs:
```diff
@@ -33,6 +36,25 @@ pub struct TransactionCost {
+// ANCHOR: time_parameters
+pub struct TimeParameters {
+    // The time to set on the first block
+    pub start_time: u64,
```
1. [`Instant`](https://doc.rust-lang.org/std/time/struct.Instant.html) is an opaque measurement, you can only [compare them](https://doc.rust-lang.org/std/time/struct.Instant.html#impl-PartialOrd%3CInstant%3E-for-Instant) to each other or check [time elapsed](https://doc.rust-lang.org/std/time/struct.Instant.html#method.elapsed) since then. Thus, there's no way to create an `Instant` representing some time point in the past.
2. If you can't compare `Instant`s to fixed values, how will `#[test] can_set_custom_block_time()` work? Is that desirable/OK?
3. If so, can `TimeParameters.block_time_interval` be a [`Duration`](https://doc.rust-lang.org/std/time/struct.Duration.html) (which can be added/subtracted to `Instant`s)?

segfault-magnet replied:
:O didn't know that. Bummer. But I'm definitely for using something that unambiguously tells when something happened. On mobile, not sure if the stdlib has something other than Instant... Still can't believe you can't instantiate a specific point in time.

Br1ght0ne replied:
Yup, we need to use something like `chrono` (or `time` for that matter) to construct `DateTime`s or `Timestamp`s. Maybe worthwhile to add it now?

MujkicA commented:
I've added `chrono` for the timestamp and removed the `TransactionResponse` wrapper to avoid introducing a breaking change in this PR. The plan is to make a new PR for it which will also include wrappers for Resource, ChainInfo etc.

digorithm commented on packages/fuels-signers/src/provider.rs:
```diff
@@ -33,6 +35,25 @@ pub struct TransactionCost {
+    // The time to set on the first block
+    pub start_time: u64,
+    // The time interval between subsequent blocks
+    pub block_time_interval: u64,
```
One thing that's not clear to me (from a user perspective) is the unit of time in this `block_time_interval`. Is it seconds, ms, ns? I think we should make it clear in the name of the property itself or at least in the comment.

MujkicA replied:
I've changed these fields to use `chrono::DateTime` and `chrono::Duration`. That should make it more transparent.

digorithm reviewed (dismissed):
Looking really good to me! Very excited to get this out :D 

W.r.t to creating wrappers; I've been divided about it, but leaning towards creating those wrappers, as the underlying structs can (and likely will, more on that later) change, so we could reduce the impact on the app devs using the SDK.

hal3e reviewed (approved):
LGTM!

Br1ght0ne reviewed (approved):
:shipit:

digorithm reviewed (approved):
Good stuff, :shipit:
